### PR TITLE
Add support for Windows managed nodegroups

### PR DESCRIPTION
**Pull request by jiechen0826**
### Description

Adds support for Windows as a node AMI family.
Allows creation of Windows nodes in managed nodegroups by setting amiFamily to one of WindowsServer2019CoreContainer, WindowsServer2019FullContainer, WindowsServer2022CoreContainer, WindowsServer2022FullContainer.

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/nodebootstrap/managed_windowss_test.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		clusterConfig := api.NewClusterConfig()
+		clusterConfig.Metadata.Name = "windohs"
+		clusterConfig.Status = &api.ClusterStatus{
+			Endpoint:                 "https://test.com",
+			CertificateAuthorityData: []byte("test"),
+		}
+
+		ng := api.NewManagedNodeGroup()
+		ng.AMIFamily = api.NodeImageFamilyWindowsServer2019CoreContainer
+		//api.SetManagedNodeGroupDefaults(ng, clusterConfig.Metadata, false)
```
Comment (cPu1) [inline pkg/nodebootstrap/managed_windowss_test.go:30]: Did you mean to comment this?

File: pkg/nodebootstrap/managed_windowss_test.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			e.updateNodeGroup(ng)
+		}
+
+		bootstrapper := nodebootstrap.NewWindowsBootstrapper(clusterConfig, ng)
+		userData, err := bootstrapper.UserData()
+		Expect(err).NotTo(HaveOccurred())
+		actual, err := base64.StdEncoding.DecodeString(userData)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(string(actual)).To(Equal(strings.TrimSpace(e.expectedUserData)))
+		//Expect(decodeData(userData)).To(Equal(strings.TrimSpace(e.expectedUserData)))
```
Comment (cPu1) [inline pkg/nodebootstrap/managed_windowss_test.go:41]: As you noticed, you can use `decodeData` and remove the `DecodeString` call above, but either approach is fine.

**Review by cPu1 — COMMENTED** [review]
I have left some suggestions, but otherwise it LGTM! The integration tests for managed nodegroup will fail though because the assertions for Windows are invalidated by this PR. I am happy to push the changes myself but looks like you'll need to give me push access to your repo and branch. Can you do that? I have the changes ready.

File: pkg/nodebootstrap/managed_windowss_test.go (hunk 2 above)
Comment (jiechen0826) [reply pkg/nodebootstrap/managed_windowss_test.go:41]: Sorry for leaving the comment there, I forgot to remove it.

File: pkg/nodebootstrap/managed_windowss_test.go (hunk 1 above)
Comment (jiechen0826) [reply pkg/nodebootstrap/managed_windowss_test.go:30]: Yes I did, I should've removed it before the PR.

File: pkg/apis/eksctl.io/v1alpha5/validation.go [hunk 3]
```diff
@@ -1037,7 +1037,9 @@ func validateNodeGroupIAM(iam *NodeGroupIAM, value, fieldName, path string) erro
 func ValidateManagedNodeGroup(index int, ng *ManagedNodeGroup) error {
 	normalizeAMIFamily(ng.BaseNodeGroup())
 	switch ng.AMIFamily {
-	case NodeImageFamilyAmazonLinux2, NodeImageFamilyBottlerocket, NodeImageFamilyUbuntu1804, NodeImageFamilyUbuntu2004:
+	case NodeImageFamilyAmazonLinux2, NodeImageFamilyBottlerocket, NodeImageFamilyUbuntu1804, NodeImageFamilyUbuntu2004, //
+		NodeImageFamilyWindowsServer2019CoreContainer, NodeImageFamilyWindowsServer2019FullContainer, //
+		NodeImageFamilyWindowsServer2022CoreContainer, NodeImageFamilyWindowsServer2022FullContainer:
 	default:
 		return errors.Errorf("%q is not supported for managed nodegroups", ng.AMIFamily)
 	}
```
[the remark this answers is not in the document]
Comment (jiechen0826) [reply pkg/apis/eksctl.io/v1alpha5/validation.go:1045]: That's a good call! I wasn't entirely sure if the case captured all the AMI Family.

File: pkg/nodebootstrap/windows.go [hunk 4]
```diff
@@ -35,8 +35,8 @@ func (b *Windows) UserData() (string, error) {
 		`<powershell>
 [string]$EKSBootstrapScriptFile = "$env:ProgramFiles\Amazon\EKS\Start-EKSBootstrap.ps1"`,
 	}
-
-	bootstrapCommands = append(bootstrapCommands, b.ng.PreBootstrapCommands...)
+	ng := (b.np).BaseNodeGroup()
```
[the remark this answers is not in the document]
Comment (jiechen0826) [reply pkg/nodebootstrap/windows.go:38]: Thank you for catching it!

File: pkg/nodebootstrap/managed_windowss_test.go (hunk 1 above)
Comment (jiechen0826) [reply pkg/nodebootstrap/managed_windowss_test.go:30]: Also I wanted to ask for your suggestion on this file name. As you can see, I named the file managed_windowss_test.go. Without the extra 's' at the end of 'windowss' the file was not picked up by the compiler due to OS mismatch. Is there other way to get around this? Or is this file name an acceptable name?

Comment (cPu1) [reply pkg/nodebootstrap/managed_windowss_test.go]: I did notice that. I think the current name has potential to be misunderstood as a typo and someone might rename it in the future. For now, we can work around this by renaming it to `windows_managed_test.go`. That way, the filename wouldn't match the `*_GOOS` build constraint. What do you think?

Comment (jiechen0826) [reply pkg/nodebootstrap/managed_windowss_test.go]: That works! Thanks for the suggestion.

File: integration/tests/windows_managed/windows_managed_test.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	Context("Create Windows managed nodegroups and run pods then cleanup", func() {
+		DescribeTable("it should be able to run Windows pods", func(windowsAmiType, workload, clusterName string) {
+			createClusterWithWindowsMNG(clusterName, windowsAmiType)
+			runWindowsPod(workload)
+			deleteCluster(clusterName)
+		},
+			Entry("Windows Server 2019 Full", api.NodeImageFamilyWindowsServer2019CoreContainer, "windows-server-iis.yaml", "managed-windows-cluster1"),
+			Entry("Windows Server 2022 Full", api.NodeImageFamilyWindowsServer2022FullContainer, "windows-server-iis-2022.yaml", "managed-windows-cluster2"),
+		)
+	})
```
Comment (cPu1) [inline integration/tests/windows_managed/windows_managed_test.go:194]: I think we don't need to have a test for creating two more clusters with different AMI families. We are wary of creating additional clusters in our integration tests as it slows them down. We can add the `runWindowsPod` call to the tests above after the nodegroup is created, and remove this entire `Context` block. What do you think?

Comment (jiechen0826) [reply integration/tests/windows_managed/windows_managed_test.go:194]: That makes sense. I will make another commit to remove the entire block. Also the lint check failed a few times, I was wondering if you have any suggestion on how to fix that?

Comment (cPu1) [reply integration/tests/windows_managed/windows_managed_test.go:194]: Can you modify the `create cluster` call in `BeforeSuite` to include Windows nodegroups by using a config file? That'll help us test `runWindowsPod`.

File: pkg/eks/nodegroup_service.go [hunk 6]
```diff
@@ -59,7 +60,7 @@ func (n *NodeGroupService) Normalize(ctx context.Context, nodePools []api.NodePo
 			if ng.LaunchTemplate == nil && ng.InstanceType == "" && len(ng.InstanceTypes) == 0 && ng.InstanceSelector.IsZero() {
 				ng.InstanceType = api.DefaultNodeType
 			}
-			hasNativeAMIFamilySupport := ng.AMIFamily == api.NodeImageFamilyAmazonLinux2 || ng.AMIFamily == api.NodeImageFamilyBottlerocket
+			hasNativeAMIFamilySupport := ng.AMIFamily == api.NodeImageFamilyAmazonLinux2 || ng.AMIFamily == api.NodeImageFamilyBottlerocket || isWindowsAMIFamily(ng.AMIFamily)
```
Comment (cPu1) [inline pkg/eks/nodegroup_service.go:63]: Any reason why `IsWindowsImage` from the `pkg/apis/eksctl.io/v1alpha5` is not being used here?
```suggestion
			hasNativeAMIFamilySupport := ng.AMIFamily == api.NodeImageFamilyAmazonLinux2 || ng.AMIFamily == api.NodeImageFamilyBottlerocket || api.IsWindowsImage(ng.AMIFamily)
```

File: pkg/eks/nodegroup_service.go [hunk 7]
```diff
@@ -273,3 +274,15 @@ func ValidateExistingNodeGroupsForCompatibility(ctx context.Context, cfg *api.Cl
...
+	case api.NodeImageFamilyWindowsServer2019CoreContainer,
+		api.NodeImageFamilyWindowsServer2019FullContainer,
+		api.NodeImageFamilyWindowsServer2022CoreContainer,
+		api.NodeImageFamilyWindowsServer2022FullContainer:
+		return true
+
+	default:
+		return false
+	}
+}
```
Comment (cPu1) [inline pkg/eks/nodegroup_service.go:288]: This function already exists in `validation.go` and can be imported here without introducing circular imports. 
```suggestion
```

Comment (cPu1) [inline pkg/eks/nodegroup_service.go:288]: This function already exists in `validation.go` and can be imported here without introducing an import cycle. 
```suggestion
```

**Comment by cPu1** [comment]
Integration tests for Windows managed nodegroups passed:

```
Ran 3 of 3 Specs in 2894.286 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 0 Skipped
```

File: pkg/eks/nodegroup_service.go (hunk 7 above)
Comment (jiechen0826) [reply pkg/eks/nodegroup_service.go]: Thank you for making the change!

File: pkg/eks/nodegroup_service.go (hunk 6 above)
Comment (jiechen0826) [reply pkg/eks/nodegroup_service.go]: I didn't realize I could import it... Thanks for fixing it!

File: pkg/nodebootstrap/windows.go [hunk 8]
```diff
@@ -35,8 +35,8 @@ func (b *Windows) UserData() (string, error) {
...
+	bootstrapCommands = append(bootstrapCommands, ng.PreBootstrapCommands...)
```
Comment (jiechen0826) [inline pkg/nodebootstrap/windows.go:40]: Hi @cPu1, is PreBootstrapCommands a valid use case for MNG?

Comment (cPu1) [reply pkg/nodebootstrap/windows.go:40]: It's supported for self-managed Windows, so I think we should support it.

**Review by cPu1 — APPROVED** [review]

**Pull request merged**